---
title: "Accurate identification of the tumor-infiltrated T cell heterogeneity in cancer immunology study"
output:
  html_document:
    theme: united
  pdf_document: default
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  # tidy = TRUE,
  # tidy.opts = list(width.cutoff = 95),
  fig.width = 10,
  message = FALSE,
  warning = FALSE,
  eval = TRUE,
  time_it = TRUE
)
```

In recent years, single-cell RNA-sequencing (scRNA-seq) enabled unbiased exploration of T cell diversity in health, disease and response to therapies at an unprecedented scale and resolution. However, a comprehensive definition of T cell “reference” subtypes remains elusive [<sup>1</sup>](#refer-anchor-1). 

Several groups try to integrate T cells in the tumor micro-environment (TME) across different cancer types to build a pan-cancer T cell landscape.  For example, Zheng et al. depicted a pan-cancer landscape of T cell heterogeneity in the TME and established a baseline reference for future scRNA-seq studies associated with cancer treatments [<sup>2</sup>](#refer-anchor-2).

In this tutorial, we demonstrate the process of projecting query T cells onto the pan-cancer T cell landscape (PTCL) to interpret the disparities in CD8 T cell heterogeneity between responsive and non-responsive patients with breast cancer who have undergone immune checkpoint blockade (ICB) therapy [<sup>3</sup>](#refer-anchor-3).  This will be done using ProjectSVR.

## Download Related Dataset

```{r load_packages, results='hide'}
library(Seurat)
library(ProjectSVR)
library(tidyverse)
options(timeout = max(3600, getOption("timeout")))
```

```{r download_dataset, eval=FALSE}
# reference model
if (!dir.exists("models")) dir.create("models")
download.file(url = "https://zenodo.org/record/8350732/files/model.Zheng2021.CD8Tcell.rds", 
              destfile = "models/model.Zheng2021.CD8Tcell.rds")
# query data
if (!dir.exists("query")) dir.create("query")
download.file(url = "https://zenodo.org/record/8350748/files/ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds", 
              destfile = "query/ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds")
```

## Map Query to Reference

### Reference mapping

```{r reference_mapping, fig.width=12, fig.height=5}
reference <- readRDS("models/model.Zheng2021.CD8Tcell.rds")
seu.q <- readRDS("query/ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds")

annotations <- c(
  "E" = "Responders",
  "n/a" = "n/a",
  "NE" = "Non-Responders"
)
seu.q$group <- annotations[seu.q$expansion]

seu.q <- ProjectSVR::MapQuery(seu.q, reference = reference, 
                              add.map.qual = T, ncores = 10)

p1 <- DimPlot(seu.q, reduction = "ref.umap", group.by = "patient_id")
p2 <- DimPlot(seu.q, reduction = "ref.umap", group.by = "cellSubType", 
              label = T) + ggsci::scale_color_d3("category20")
p1 + p2
```

### Maping quality

Determine the threshold to distinguish the good and bad projections.

QC plot for the reference mapping.

```{r map_qc_plot, fig.width=14, fig.height=4}
## cutoff by adjusted p value
MapQCPlot(seu.q, p.adj.cutoff = 1e-3)
## or mean.knn.dist
MapQCPlot(seu.q, map.q.cutoff = 1.6)
```

Visualize the projected query cells onto the reference atlas.

```{r projection_plot, fig.width=15, fig.height=8}
PlotProjection(seu.q, reference, split.by = "cellSubType", ref.color.by = "cluster.name",
               ref.size = .5, ref.alpha = .3, query.size = 1, query.alpha = .5, n.row = 2)
```

### Label transfer

```{r label_transfer, fig.width=10, fig.height=4}
seu.q <- ProjectSVR::LabelTransfer(seu.q, reference, ref.label.col = "cluster.name")
```

## Differential cell population

```{r, fig.width=11, fig.height=5}
seu.q <- subset(seu.q, group != "n/a" & timepoint == "Pre")
seu.q <- subset(seu.q, mean.knn.dist < 1.6)
seu.q$group <- factor(seu.q$group)

PlotProjection(seu.q, reference, split.by = "group", ref.color.by = "cluster.name",
               ref.size = .5, ref.alpha = .3, query.size = 1, query.alpha = .5, n.row = 1)
```


### Alluvia plot

```{r alluvia_plot, fig.width=6, fig.height=5}
AlluviaPlot(seu.q@meta.data, by = "group", fill = "knn.pred.celltype",
            colors = reference$ref.cellmeta$colors,
            bar.width = .5)
```

### Preference analysis

```{r preference_plot, fig.width=6, fig.height=5}
GroupPreferencePlot(seu.q@meta.data, group.by = "group", 
                    preference.on = "knn.pred.celltype", 
                    column_names_rot = 0, column_names_centered = TRUE)
```

If `seu.q$knn.pred.celltype` was not a factor, then rows were ordered by o/e value.

```{r preference_plot2, fig.width=6, fig.height=5}
meta.data <- seu.q@meta.data
meta.data$knn.pred.celltype <- as.character(meta.data$knn.pred.celltype)
GroupPreferencePlot(meta.data, group.by = "group", 
                    preference.on = "knn.pred.celltype", 
                    column_names_rot = 0, column_names_centered = TRUE)
```

### Wilcoxon test

```{r wilcoxon_test, fig.width=5, fig.height=4}
da.test <- AbundanceTest(cellmeta = seu.q@meta.data,
                         celltype.col = "knn.pred.celltype",
                         sample.col = "patient_id",
                         group.col = "group")

## Volcano plot
VolcanoPlot(da.test, xlab = "log2(R / NR)", ylab = "Frequency",
            colors = reference$ref.cellmeta$colors)
```

Focus on the detailed information via `BoxPlot()`.

```{r box_plot, fig.width=6, fig.height=5}
## Box plot
BoxPlot(cellmeta = seu.q@meta.data, celltype.col = "knn.pred.celltype",
        sample.col = "patient_id", group.col = "group",
        celltypes.show = c("CD8.c12(terminal Tex)", "CD8.c11(GZMK+ Tex)"),
        legend.ncol = 2) +
  scale_x_discrete(labels = c("NR", "R"))
```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>


## References

<div id="refer-anchor-1"></div>
- [1] [Andreatta M, et al. Interpretation of T cell states from single-cell transcriptomics data using reference atlases. Nat Commun. 2021 May 20;12(1):2965. doi: 10.1038/s41467-021-23324-4. PMID: 34017005.](https://pubmed.ncbi.nlm.nih.gov/34017005/)

<div id="refer-anchor-2"></div>
- [2] [Zheng L, et al. Pan-cancer single-cell landscape of tumor-infiltrating T cells. Science. 2021 Dec 17;374(6574):abe6474. doi: 10.1126/science.abe6474. Epub 2021 Dec 17. PMID: 34914499.](https://pubmed.ncbi.nlm.nih.gov/34914499/)

<div id="refer-anchor-3"></div>
- [3] [Bassez A, et al. A single-cell map of intratumoral changes during anti-PD1 treatment of patients with breast cancer. Nat Med. 2021 May;27(5):820-832. doi: 10.1038/s41591-021-01323-8. Epub 2021 May 6. PMID: 33958794.](https://pubmed.ncbi.nlm.nih.gov/33958794/)
