<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>NA • ProjectSVR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css"><script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="pkgdown.js"></script><meta property="og:title" content="NA"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-W1XR84YEPZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-W1XR84YEPZ');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-title-body">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ProjectSVR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Generic example</li>
    <li>
      <a href="articles/quick_start.html">Quick start</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Build reference model</li>
    <li>
      <a href="articles/model_pbmc.html">DISCO blood atlas</a>
    </li>
    <li>
      <a href="articles/model_mfia.html">Maternal-fetal interface atlas</a>
    </li>
    <li>
      <a href="articles/model_pctl.html">Pan-cancer T cell lanscape</a>
    </li>
    <li>
      <a href="articles/model_mtca.html">Mouse testicular cell atlas</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Map query dataset</li>
    <li>
      <a href="articles/mapQuery_immune_RPL.html">Decidual immune microenvironment of RPL patients</a>
    </li>
    <li>
      <a href="articles/mapQuery_ICB_BRCA.html">ICB responsiveness of BRCA</a>
    </li>
    <li>
      <a href="articles/mapQuery_Zfp541-KO.html">Genetic perturbed germ cells (Zfp541-KO)</a>
    </li>
    <li>
      <a href="articles/mapQuery_Ythdc2-KO.html">Genetic perturbed germ cells (Ythdc2-KO)</a>
    </li>
    <li>
      <a href="articles/mapQuery_invitro_meiosis.html">In vitro induced meiosis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">MISC</li>
    <li>
      <a href="articles/misc_Ythdc2-KO_pseudotime.html">Training the reference pseudotime model</a>
    </li>
    <li>
      <a href="articles/misc_disco_pbmc_svm.html">Training the SVM model for cell type prediction</a>
    </li>
  </ul></li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/JarningGau/ProjectSVR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>NA</h1>
    </div>

<div class="section level2">
<h2 id="log">Log<a class="anchor" aria-label="anchor" href="#log"></a></h2>
<ul><li>
<p><code>2022-07-12</code></p>
<p><code>TODO</code> <code>Enhancement</code> Try to speed up <code><a href="reference/MergeCells.html">MergeCells()</a></code> function.</p>
</li>
<li>
<p><code>2022-07-11</code></p>
<p><code>Done</code> import <code>mlr3verse</code> properly</p>
<p><code>Done</code> Add normalization method for <code>FitEnsemblSVM()</code> <code><a href="reference/FitEnsemblMultiClassif.html">FitEnsemblMultiClassif()</a></code> <code><a href="reference/PredictNewdata.html">PredictNewdata()</a></code> and <code><a href="reference/ProjectNewdata.html">ProjectNewdata()</a></code></p>
</li>
<li>
<p><code>2022-06-11</code></p>
<p><code>Enhancement</code> Plot for the VD results.</p>
</li>
<li>
<p><code>2022-06-09</code></p>
<p><code>Done</code> <code>Enhancement</code> Normalize the feature.mat by cells (sum to 1) before model construction and prediction.</p>
</li>
<li>
<p><code>2022-05-30</code></p>
<p><code>Done</code> <code>BUG</code> FitEnsembleSVM() | feature.mat should be a data.frame, when input a matrix. error occurs. Want a clear error information.</p>
</li>
<li>
<p><code>2022-05-28</code></p>
<p><code>Enhancement</code> Two prediction function is too confused for new users. Try to merge them via generic method.</p>
<p><code>Enhancement</code> Add a plot function for visualizing the coords before and after refinement.</p>
<p><code>Done</code> <code>BUG</code> AddProjQual() function will generated duplicated results in @cellmeta slot when execute twice or more times. Want codes for colnames check.</p>
</li>
<li>
<p><code>2022-05-24</code></p>
<p><code>Done</code> <code><a href="reference/LabelTransfer.html">LabelTransfer()</a></code> SVM model.</p>
<p>​ <code><a href="reference/FitEnsemblMultiClassif.html">FitEnsemblMultiClassif()</a></code></p>
<p>​ <code><a href="reference/PredictNewdata.html">PredictNewdata()</a></code></p>
<p>​ <code><a href="reference/MajorityVote.html">MajorityVote()</a></code></p>
<p><code>Done</code> Test these functions.</p>
</li>
<li>
<p><code>2022-05-23</code></p>
<p><code>Done</code> Differential expression analysis <code>VarDecompose()</code> and <code>GetExpressionShift()</code></p>
<p><code>TODO</code> <code>ClusterGenes()</code></p>
<p><code>Done</code> Test <code>VarDecompose()</code> and <code>GetExpressionShift()</code></p>
</li>
<li>
<p><code>2022-05-19</code></p>
<p><code>Done</code> Add <code>CellProject</code> object to store results generated during cell projection.</p>
<p><code>Done</code> New function: <code>RefineProjection</code></p>
</li>
<li>
<p><code>2022-04-30</code></p>
<p><code>Done</code> <code><a href="reference/ProjectNewdata.html">ProjectNewdata()</a></code> add warning info and fill up 0 for not enough feature case. (case2 Ythdc2-KO data)</p>
<p><del><code>TODO</code> Add a function to refine low confidence mapping coords.</del></p>
<p><del><code>TODO</code> <code><a href="reference/LabelTransfer.html">LabelTransfer()</a></code> Add SVM method.</del></p>
</li>
<li>
<p><code>2022-03-27</code></p>
<p><code>Done</code> ggplot2 in scProject package. <code>import ggplot2</code> is not proper for a package.</p>
</li>
<li>
<p><code>2022-03-26</code></p>
<p><code>TODO</code> Add new signature scoring method – <a href="https://github.com/carmonalab/UCell" class="external-link">UCell</a>.</p>
<p><code>Done</code> RNA velocity on <code>scenario 2</code> task</p>
</li>
<li>
<p><code>2022-03-25</code></p>
<p><del><code>TODO</code> Differential expressed genes. <code>cacoa</code> package solved the problem.</del></p>
</li>
<li>
<p><code>2022-03-24</code></p>
<p><code>Done</code> Optimize the mapQ function:</p>
<ol style="list-style-type: decimal"><li>Markov affinity matrix replaces the distance matrix to calculate the PCCs. <code>Done</code> <code>Not good</code>
</li>
<li>NN-based mean distance performs better. <code>Done</code>
</li>
</ol></li>
<li>
<p><code>2022-03-23</code></p>
<p><code>Done</code> mTCA, levels’ cellular annotation.</p>
</li>
<li>
<p><code>2022-03-22</code></p>
<p><code>Done</code> Add two mapQ functions</p>
<ol style="list-style-type: decimal"><li>Global (random N) rather than local (KNN) topological similarity</li>
<li>Jaccard index</li>
</ol></li>
<li>
<p><code>2022-03-21</code></p>
<p>scenario2: time-course gonocyte-spermatogonia-transition</p>
<p><code>Done</code> optimized model training step.</p>
<ol style="list-style-type: decimal"><li>Drop outliers cells before model training; <code>TODO</code>
</li>
<li>redo cNMF on cluster-based rather than grid-based pseudocells; <code>Done</code>
</li>
<li>The number of cells from different cell type should be balanced (balanced sampling replaces random sampling). <code>TODO</code>
</li>
<li>Only use protein coding genes. <code>Done</code> <code>Boost the results</code>
</li>
</ol></li>
<li>
<p><code>2022-03-20</code></p>
<p><code>Done</code> Test <code>MakeMetaCells()</code></p>
<p><code>TODO</code> Drop the <code>ml3verse</code> dependence. Use <code>e1071</code> for SVR directly.</p>
</li>
<li>
<p><code>2022-03-18</code></p>
<p>Speed up <code><a href="reference/MergeCells.html">MergeCells()</a></code></p>
<p>Add <code><a href="reference/LabelTransfer.html">LabelTransfer()</a></code> for cell type annotation after projection.</p>
<p>Add <code>MakeMetaCell()</code> function for group transcriptome similar cells into meta cells to speed up cNMF.</p>
<p><code>TODO</code> Add <code>AddProjQaul()</code>, cluster-based map quality.</p>
<p><code>TODO</code> Implementation of DE analysis model.</p>
</li>
<li>
<p><code>2022-03-17</code></p>
<p>Add <code><a href="reference/NNHelper.html">NNHelper()</a></code> function for calculating nearest neightbors.</p>
<p>Modifiy <code><a href="reference/EstimateKnnDensity.html">EstimateKnnDensity()</a></code>. Test it again. Drop @param cores.</p>
<p>Implementation of <code><a href="reference/AddProjQual.html">AddProjQual()</a></code>, evaluating projection/mapping quality.</p>
</li>
<li>
<p><code>2022-03-15</code></p>
<p>test cNMF.R worked</p>
<p>implementation of <code><a href="reference/FitEnsembleSVM.html">FitEnsembleSVM()</a></code>, <code><a href="reference/ComputeModuleScore.html">ComputeModuleScore()</a></code></p>
</li>
<li>
<p><code>2022-03-14</code></p>
<p><code>add function</code> merge by bins, or merge by mesh points.</p>
</li>
<li>
<p><code>2022-03-12</code></p>
<p><code><a href="reference/EstimateKnnDensity.html">EstimateKnnDensity()</a></code>注意数字在R中默认为<code>float</code>，这在利用reticulate和python交互时容易出现类型错误</p>
<p>函数中需要一个类型检查。</p>
</li>
</ul></div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Jianing Gao.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

