<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Accurate identification of the tumor-infiltrated T cell heterogeneity in cancer immunology study • ProjectSVR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Accurate identification of the tumor-infiltrated T cell heterogeneity in cancer immunology study"><meta property="og:description" content="ProjectSVR"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-W1XR84YEPZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-W1XR84YEPZ');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ProjectSVR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/quick_start.html">Quick start</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/JarningGau/ProjectSVR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>





<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate identification of the tumor-infiltrated T cell heterogeneity in cancer immunology study</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 21, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/JarningGau/ProjectSVR/blob/HEAD/vignettes/ICB_breast_cancer.Rmd" class="external-link"><code>vignettes/ICB_breast_cancer.Rmd</code></a></small>
      <div class="hidden name"><code>ICB_breast_cancer.Rmd</code></div>

    </div>

    
    
<p>In recent years, single-cell RNA-sequencing (scRNA-seq) enabled unbiased exploration of T cell diversity in health, disease and response to therapies at an unprecedented scale and resolution. However, a comprehensive definition of T cell “reference” subtypes remains elusive <a href="#refer-anchor-1"><sup>1</sup></a>.</p>
<p>Several groups try to integrate T cells in the tumor micro-environment (TME) across different cancer types to build a pan-cancer T cell landscape. For example, Zheng et al. depicted a pan-cancer landscape of T cell heterogeneity in the TME and established a baseline reference for future scRNA-seq studies associated with cancer treatments <a href="#refer-anchor-2"><sup>2</sup></a>.</p>
<p>In this tutorial, we demonstrate the process of projecting query T cells onto the pan-cancer T cell landscape (PTCL) to interpret the disparities in CD8 T cell heterogeneity between responsive and non-responsive patients with breast cancer who have undergone immune checkpoint blockade (ICB) therapy <a href="#refer-anchor-3"><sup>3</sup></a>. This will be done using ProjectSVR.</p>
<div id="download-related-dataset" class="section level2">
<h2>Download related dataset</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ProjectSVR)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="fu">max</span>(<span class="dv">3600</span>, <span class="fu">getOption</span>(<span class="st">&quot;timeout&quot;</span>)))</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reference data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We provide reference in qs format for quicklyu saving and readling objects to and from disk.</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># https://cran.r-project.org/web/packages/qs/vignettes/vignette.html</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">&quot;&quot;</span>, <span class="at">destfile =</span> <span class="st">&quot;ZhengLiangtao.CD8.seurat.qs&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># query data</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">&quot;&quot;</span>, <span class="at">destfile =</span> <span class="st">&quot;ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds&quot;&quot;)</span></span></code></pre></div>
</div>
<div id="build-reference-model" class="section level2">
<h2>Build Reference Model</h2>
<p>Load reference dataset. See also</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>seu.ref <span class="ot">&lt;-</span> qs<span class="sc">::</span><span class="fu">qread</span>(<span class="st">&quot;ZhengLiangtao.CD8.seurat.qs&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>data.plot <span class="ot">&lt;-</span> seu.ref<span class="sc">@</span>misc<span class="sc">$</span>data.refplot<span class="sc">$</span>meta.data</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data.plot, <span class="fu">aes</span>(UMAP_1, UMAP_2, <span class="at">color =</span> cluster.name), <span class="at">pt.size =</span> .<span class="dv">4</span>) <span class="sc">+</span> </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> seu.ref<span class="sc">@</span>misc<span class="sc">$</span>data.refplot<span class="sc">$</span>colors) <span class="sc">+</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">inherit.aes =</span> F, <span class="at">data =</span> seu.ref<span class="sc">@</span>misc<span class="sc">$</span>data.refplot<span class="sc">$</span>text.pos,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">mapping =</span> <span class="fu">aes</span>(x, y, <span class="at">label =</span> label), <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span> </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">2</span>))) <span class="sc">+</span> </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">ratio =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/referenc_plot-1.png" width="672" /> ### Transfer raw count matrix to gene set score matrix</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>top.genes <span class="ot">&lt;-</span> seu.ref<span class="sc">@</span>misc<span class="sc">$</span>markers</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>top.genes <span class="ot">&lt;-</span> <span class="fu">lapply</span>(top.genes, <span class="cf">function</span>(xx) <span class="fu">head</span>(xx, <span class="dv">20</span>))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(top.genes, length)</span></code></pre></div>
<pre><code>## CD8.c01 CD8.c02 CD8.c03 CD8.c04 CD8.c05 CD8.c06 CD8.c07 CD8.c08 CD8.c09 CD8.c10 
##      20      20       7      20      20      20      20      20      20      20 
## CD8.c11 CD8.c12 CD8.c13 CD8.c14 CD8.c15 CD8.c16 CD8.c17 
##      20      20      20      20      20      20      20</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>seu.ref <span class="ot">&lt;-</span> <span class="fu">ComputeModuleScore</span>(seu.ref, <span class="at">gene.sets =</span> top.genes, <span class="at">method =</span> <span class="st">&quot;UCell&quot;</span>, <span class="at">cores =</span> <span class="dv">20</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The signature score matrix is stored in &#39;SignatureScore&#39; assay</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">Assays</span>(seu.ref)</span></code></pre></div>
<pre><code>## [1] &quot;RNA&quot;            &quot;SignatureScore&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seu.ref) <span class="ot">&lt;-</span> <span class="st">&quot;SignatureScore&quot;</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seu.ref, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;CD8.c12&quot;</span>, <span class="st">&quot;CD8.c16&quot;</span>), <span class="at">pt.size =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/add_gene_module_score-1.png" width="768" /></p>
<div id="training-reference-model" class="section level3">
<h3>Training reference model</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>gss.mat <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seu.ref, <span class="at">vars =</span> <span class="fu">rownames</span>(seu.ref))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>embeddings.df <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seu.ref, <span class="at">vars =</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>batch.size <span class="ot">=</span> <span class="dv">8000</span> <span class="co"># number of subsampled cells for each SVR model </span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>n.models <span class="ot">=</span> <span class="dv">10</span>      <span class="co"># number of SVR models trained</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>umap.model <span class="ot">&lt;-</span> <span class="fu">FitEnsembleSVM</span>(<span class="at">feature.mat =</span> gss.mat,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>                             <span class="at">emb.mat =</span> embeddings.df,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>                             <span class="at">batch.size =</span> batch.size,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>                             <span class="at">n.models =</span> n.models,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>                             <span class="at">cores =</span> <span class="dv">10</span>)</span></code></pre></div>
</div>
<div id="save-the-reference-model" class="section level3">
<h3>Save the reference model</h3>
<p>ref.cellmeta stores:</p>
<ol style="list-style-type: decimal">
<li><p>[optional] colors: for plots</p></li>
<li><p>[optional] text.pos: text annotation on the reference plots</p></li>
<li><p>meta.data: cell meta data (embeddings + cell type information)</p></li>
</ol>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>ref.cellmeta <span class="ot">&lt;-</span> seu.ref<span class="sc">@</span>misc<span class="sc">$</span>data.refplot</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>bg.genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(seu.ref[[<span class="st">&quot;RNA&quot;</span>]])</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;models&quot;</span> <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;umap&quot;</span> <span class="ot">=</span> umap.model</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;genes&quot;</span> <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;gene.sets&quot;</span> <span class="ot">=</span> top.genes, <span class="co"># list</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;bg.genes&quot;</span> <span class="ot">=</span> bg.genes <span class="co"># vector</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;ref.cellmeta&quot;</span> <span class="ot">=</span> ref.cellmeta <span class="co"># list for reference plot</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(reference, <span class="st">&quot;model.Zheng.CD8Tcell.rds&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="map-query-to-reference" class="section level2">
<h2>Map Query to Reference</h2>
<div id="reference-mapping" class="section level3">
<h3>Reference mapping</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;model.Zheng.CD8Tcell.rds&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>seu.q <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;E&quot;</span> <span class="ot">=</span> <span class="st">&quot;Responders&quot;</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;n/a&quot;</span> <span class="ot">=</span> <span class="st">&quot;n/a&quot;</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;NE&quot;</span> <span class="ot">=</span> <span class="st">&quot;Non-Responders&quot;</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>seu.q<span class="sc">$</span>group <span class="ot">&lt;-</span> annotations[seu.q<span class="sc">$</span>expansion]</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>seu.q <span class="ot">&lt;-</span> ProjectSVR<span class="sc">::</span><span class="fu">MapQuery</span>(seu.q, <span class="at">reference =</span> reference, </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>                              <span class="at">add.map.qual =</span> T, <span class="at">ncores =</span> <span class="dv">10</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seu.q, <span class="at">reduction =</span> <span class="st">&quot;ref.umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;patient_id&quot;</span>)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seu.q, <span class="at">reduction =</span> <span class="st">&quot;ref.umap&quot;</span>, <span class="at">group.by =</span> <span class="st">&quot;cellSubType&quot;</span>, </span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>              <span class="at">label =</span> T) <span class="sc">+</span> ggsci<span class="sc">::</span><span class="fu">scale_color_d3</span>(<span class="st">&quot;category20&quot;</span>)</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/reference_mapping-1.png" width="1152" /></p>
</div>
<div id="maping-quality" class="section level3">
<h3>Maping quality</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>data.plot <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seu.q, <span class="at">vars =</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">&quot;refUMAP_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>), <span class="st">&quot;mean.knn.dist&quot;</span>, <span class="st">&quot;mapQ.p.adj&quot;</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data.plot, <span class="fu">aes</span>(mean.knn.dist, <span class="sc">-</span><span class="fu">log10</span>(mapQ.p.adj))) <span class="sc">+</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">1.6</span>, <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/map_qc_threshold-1.png" width="480" /></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>data.plot <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seu.q, <span class="at">vars =</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">&quot;refUMAP_&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>), <span class="st">&quot;mean.knn.dist&quot;</span>, <span class="st">&quot;mapQ.p.adj&quot;</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>cut.off <span class="ot">&lt;-</span> <span class="fl">1.6</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>low.qual.prop <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(<span class="fu">table</span>(data.plot<span class="sc">$</span>mean.knn.dist <span class="sc">&gt;</span> cut.off))[<span class="st">&quot;TRUE&quot;</span>]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>low.qual.prop <span class="ot">&lt;-</span> <span class="fu">round</span>(low.qual.prop, <span class="dv">4</span>) <span class="sc">*</span><span class="dv">100</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data.plot, <span class="fu">aes</span>(refUMAP_1, refUMAP_2)) <span class="sc">+</span> </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> mean.knn.dist), <span class="at">size =</span> .<span class="dv">2</span>) <span class="sc">+</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">colours =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="at">values =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="dv">1</span>)) <span class="sc">+</span> </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Mapping quality&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;black&quot;</span>),</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> .<span class="dv">5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>))</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data.plot, <span class="fu">aes</span>(mean.knn.dist)) <span class="sc">+</span> </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">fill =</span> <span class="st">&quot;lightblue&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> cut.off, <span class="at">linetype =</span> <span class="st">&quot;dashed&quot;</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Distribution of mean </span><span class="sc">\n</span><span class="st">kNN dist&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> </span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;black&quot;</span>),</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> .<span class="dv">5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>))</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data.plot, <span class="fu">aes</span>(refUMAP_1, refUMAP_2)) <span class="sc">+</span> </span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> mean.knn.dist <span class="sc">&gt;</span> cut.off, <span class="at">size =</span> mean.knn.dist <span class="sc">&gt;</span> cut.off)) <span class="sc">+</span> </span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">sprintf</span>(<span class="st">&quot;Low-quality projection</span><span class="sc">\n</span><span class="st">(%s%%)&quot;</span>, low.qual.prop)) <span class="sc">+</span> </span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span> </span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_size_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.5</span>), <span class="at">guide =</span> <span class="st">&quot;none&quot;</span>) <span class="sc">+</span> </span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="fu">sprintf</span>(<span class="st">&quot;mean.knn.dist</span><span class="sc">\n</span><span class="st">(&gt;%s)&quot;</span>, cut.off))) <span class="sc">+</span> </span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>(<span class="at">base_size =</span> <span class="dv">15</span>) <span class="sc">+</span> </span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">&quot;black&quot;</span>),</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> .<span class="dv">5</span>, <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>))</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> p3</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/map_qc_plot-1.png" width="1344" /></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">PlotProjection</span>(seu.q, reference, <span class="at">split.by =</span> <span class="st">&quot;cellSubType&quot;</span>, <span class="at">ref.color.by =</span> <span class="st">&quot;cluster.name&quot;</span>,</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">ref.size =</span> .<span class="dv">5</span>, <span class="at">ref.alpha =</span> .<span class="dv">3</span>, <span class="at">query.size =</span> <span class="dv">1</span>, <span class="at">query.alpha =</span> .<span class="dv">5</span>, <span class="at">n.row =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/projection_plot-1.png" width="1440" /></p>
</div>
<div id="label-transfer" class="section level3">
<h3>Label transfer</h3>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>seu.q <span class="ot">&lt;-</span> ProjectSVR<span class="sc">::</span><span class="fu">LabelTransfer</span>(seu.q, reference, <span class="at">ref.label.col =</span> <span class="st">&quot;cluster.name&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="differential-cell-population" class="section level2">
<h2>Differential cell population</h2>
<div id="alluvia-plot" class="section level3">
<h3>Alluvia plot</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>seu.q <span class="ot">&lt;-</span> <span class="fu">subset</span>(seu.q, group <span class="sc">!=</span> <span class="st">&quot;n/a&quot;</span> <span class="sc">&amp;</span> timepoint <span class="sc">==</span> <span class="st">&quot;Pre&quot;</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>seu.q <span class="ot">&lt;-</span> <span class="fu">subset</span>(seu.q, mean.knn.dist <span class="sc">&lt;</span> <span class="fl">1.6</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>seu.q<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(seu.q<span class="sc">$</span>group)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">AlluviaPlot</span>(seu.q<span class="sc">@</span>meta.data, <span class="at">by =</span> <span class="st">&quot;group&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;knn.pred.celltype&quot;</span>,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">colors =</span> reference<span class="sc">$</span>ref.cellmeta<span class="sc">$</span>colors,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">bar.width =</span> .<span class="dv">5</span>)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/alluvia_plot-1.png" width="576" /></p>
</div>
<div id="preference-analysis" class="section level3">
<h3>Preference analysis</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">GroupPreferencePlot</span>(seu.q<span class="sc">@</span>meta.data, <span class="at">group.by =</span> <span class="st">&quot;group&quot;</span>, </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">preference.on =</span> <span class="st">&quot;knn.pred.celltype&quot;</span>, </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">column_names_rot =</span> <span class="dv">0</span>, <span class="at">column_names_centered =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/preference_plot-1.png" width="576" /></p>
</div>
<div id="wilcoxon-test" class="section level3">
<h3>Wilcoxon test</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>da.test <span class="ot">&lt;-</span> <span class="fu">AbundanceTest</span>(<span class="at">cellmeta =</span> seu.q<span class="sc">@</span>meta.data,</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">celltype.col =</span> <span class="st">&quot;knn.pred.celltype&quot;</span>,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">sample.col =</span> <span class="st">&quot;patient_id&quot;</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">group.col =</span> <span class="st">&quot;group&quot;</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Volcano plot</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="fu">VolcanoPlot</span>(da.test, <span class="at">xlab =</span> <span class="st">&quot;log2(R / NR)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Frequency&quot;</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">colors =</span> reference<span class="sc">$</span>ref.cellmeta<span class="sc">$</span>colors)</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/wilcoxon_test-1.png" width="480" /></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Box plot</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">BoxPlot</span>(<span class="at">cellmeta =</span> seu.q<span class="sc">@</span>meta.data, <span class="at">celltype.col =</span> <span class="st">&quot;knn.pred.celltype&quot;</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">sample.col =</span> <span class="st">&quot;patient_id&quot;</span>, <span class="at">group.col =</span> <span class="st">&quot;group&quot;</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">celltypes.show =</span> <span class="fu">c</span>(<span class="st">&quot;CD8.c12(terminal Tex)&quot;</span>, <span class="st">&quot;CD8.c11(GZMK+ Tex)&quot;</span>),</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.ncol =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_discrete</span>(<span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;NR&quot;</span>, <span class="st">&quot;R&quot;</span>))</span></code></pre></div>
<p><img src="ICB_breast_cancer_files/figure-html/box_plot-1.png" width="576" /></p>
</div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<div id="refer-anchor-1">

</div>
<ul>
<li>[1] <a href="https://pubmed.ncbi.nlm.nih.gov/34017005/">Andreatta M, et al. Interpretation of T cell states from single-cell transcriptomics data using reference atlases. Nat Commun. 2021 May 20;12(1):2965. doi: 10.1038/s41467-021-23324-4. PMID: 34017005.</a></li>
</ul>
<div id="refer-anchor-2">

</div>
<ul>
<li>[2] <a href="https://pubmed.ncbi.nlm.nih.gov/34914499/">Zheng L, et al. Pan-cancer single-cell landscape of tumor-infiltrating T cells. Science. 2021 Dec 17;374(6574):abe6474. doi: 10.1126/science.abe6474. Epub 2021 Dec 17. PMID: 34914499.</a></li>
</ul>
<div id="refer-anchor-3">

</div>
<ul>
<li>[3] <a href="https://pubmed.ncbi.nlm.nih.gov/33958794/">Bassez A, et al. A single-cell map of intratumoral changes during anti-PD1 treatment of patients with breast cancer. Nat Med. 2021 May;27(5):820-832. doi: 10.1038/s41591-021-01323-8. Epub 2021 May 6. PMID: 33958794.</a></li>
</ul>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.1.2 (2021-11-01)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS:   /opt/R4.1/lib64/R/lib/libRblas.so
## LAPACK: /opt/R4.1/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] forcats_0.5.2         stringr_1.4.1         dplyr_1.0.10         
##  [4] purrr_0.3.4           readr_2.1.2           tidyr_1.2.1          
##  [7] tibble_3.1.8          ggplot2_3.4.1         tidyverse_1.3.2      
## [10] ProjectSVR_0.1.0.9000 sp_1.5-0              SeuratObject_4.1.2   
## [13] Seurat_4.2.0         
## 
## loaded via a namespace (and not attached):
##   [1] utf8_1.2.2             reticulate_1.26        tidyselect_1.1.2      
##   [4] mlr3learners_0.5.4     htmlwidgets_1.5.4      BiocParallel_1.28.3   
##   [7] grid_4.1.2             Rtsne_0.16             mlr3misc_0.11.0       
##  [10] munsell_0.5.0          codetools_0.2-18       bbotk_0.5.4           
##  [13] ragg_1.2.2             ica_1.0-3              future_1.28.0         
##  [16] miniUI_0.1.1.1         mlr3verse_0.2.5        withr_2.5.0           
##  [19] spatstat.random_3.0-1  colorspace_2.0-3       progressr_0.11.0      
##  [22] highr_0.9              knitr_1.40             uuid_1.1-0            
##  [25] rstudioapi_0.14        stats4_4.1.2           ROCR_1.0-11           
##  [28] ggsignif_0.6.3         tensor_1.5             listenv_0.8.0         
##  [31] labeling_0.4.2         mlr3tuning_0.14.0      lgr_0.4.4             
##  [34] polyclip_1.10-0        farver_2.1.1           rprojroot_2.0.3       
##  [37] parallelly_1.32.1      vctrs_0.5.1            generics_0.1.3        
##  [40] clusterCrit_1.2.8      xfun_0.33              R6_2.5.1              
##  [43] doParallel_1.0.17      clue_0.3-61            isoband_0.2.5         
##  [46] spatstat.utils_3.0-1   cachem_1.0.6           assertthat_0.2.1      
##  [49] promises_1.2.0.1       scales_1.2.1           googlesheets4_1.0.1   
##  [52] rgeos_0.5-9            gtable_0.3.1           globals_0.16.1        
##  [55] goftest_1.2-3          rlang_1.0.6            systemfonts_1.0.4     
##  [58] GlobalOptions_0.1.2    splines_4.1.2          lazyeval_0.2.2        
##  [61] gargle_1.2.1           paradox_0.10.0         spatstat.geom_3.0-3   
##  [64] broom_1.0.1            checkmate_2.1.0        modelr_0.1.9          
##  [67] yaml_2.3.5             reshape2_1.4.4         abind_1.4-5           
##  [70] mlr3_0.14.0            backports_1.4.1        httpuv_1.6.6          
##  [73] tools_4.1.2            ellipsis_0.3.2         spatstat.core_2.4-4   
##  [76] jquerylib_0.1.4        RColorBrewer_1.1-3     BiocGenerics_0.40.0   
##  [79] ggridges_0.5.3         Rcpp_1.0.9             plyr_1.8.7            
##  [82] rpart_4.1.16           deldir_1.0-6           pbapply_1.5-0         
##  [85] GetoptLong_1.0.5       cowplot_1.1.1          S4Vectors_0.32.4      
##  [88] zoo_1.8-11             haven_2.5.1            ggrepel_0.9.1         
##  [91] cluster_2.1.4          here_1.0.1             fs_1.5.2              
##  [94] magrittr_2.0.3         data.table_1.14.2      scattermore_0.8       
##  [97] circlize_0.4.15        reprex_2.0.2           lmtest_0.9-40         
## [100] RANN_2.6.1             googledrive_2.0.0      fitdistrplus_1.1-8    
## [103] matrixStats_0.62.0     stringfish_0.15.8      qs_0.25.5             
## [106] hms_1.1.2              patchwork_1.1.2        mime_0.12             
## [109] evaluate_0.16          xtable_1.8-4           readxl_1.4.1          
## [112] IRanges_2.28.0         gridExtra_2.3          shape_1.4.6           
## [115] UCell_1.3.1            compiler_4.1.2         mlr3cluster_0.1.4     
## [118] KernSmooth_2.23-20     crayon_1.5.1           htmltools_0.5.3       
## [121] tzdb_0.3.0             mgcv_1.8-40            later_1.3.0           
## [124] RcppParallel_5.1.6     RApiSerialize_0.1.2    lubridate_1.8.0       
## [127] DBI_1.1.3              dbplyr_2.2.1           ComplexHeatmap_2.10.0 
## [130] rappdirs_0.3.3         MASS_7.3-58.1          mlr3data_0.6.1        
## [133] Matrix_1.5-1           cli_3.4.1              parallel_4.1.2        
## [136] igraph_1.3.5           pkgconfig_2.0.3        pkgdown_2.0.6         
## [139] plotly_4.10.0          spatstat.sparse_3.0-0  xml2_1.3.3            
## [142] foreach_1.5.2          bslib_0.4.0            mlr3fselect_0.7.2     
## [145] rvest_1.0.3            digest_0.6.29          sctransform_0.3.5     
## [148] RcppAnnoy_0.0.19       mlr3filters_0.6.0      spatstat.data_3.0-0   
## [151] cellranger_1.1.0       rmarkdown_2.16         leiden_0.4.3          
## [154] uwot_0.1.14            curl_5.0.1             shiny_1.7.2           
## [157] rjson_0.2.21           lifecycle_1.0.3        nlme_3.1-155          
## [160] jsonlite_1.8.0         mlr3tuningspaces_0.3.0 desc_1.4.2            
## [163] viridisLite_0.4.1      fansi_1.0.3            pillar_1.8.1          
## [166] ggsci_2.9              lattice_0.20-45        fastmap_1.1.0         
## [169] httr_1.4.4             survival_3.4-0         glue_1.6.2            
## [172] mlr3viz_0.5.10         png_0.1-7              iterators_1.0.14      
## [175] stringi_1.7.6          sass_0.4.2             mlr3pipelines_0.4.2   
## [178] palmerpenguins_0.1.1   textshaping_0.3.6      memoise_2.0.1         
## [181] irlba_2.3.5            future.apply_1.9.1</code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Jianing Gao.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>
