<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate identification of the tumor-infiltrated T cell heterogeneity in cancer immunology study • ProjectSVR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accurate identification of the tumor-infiltrated T cell heterogeneity in cancer immunology study">
<meta property="og:description" content="ProjectSVR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-W1XR84YEPZ"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-W1XR84YEPZ');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ProjectSVR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Generic example</li>
    <li>
      <a href="../articles/quick_start.html">Quick start</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Build reference model</li>
    <li>
      <a href="../articles/model_pbmc.html">DISCO blood atlas</a>
    </li>
    <li>
      <a href="../articles/model_mfia.html">Maternal-fetal interface atlas</a>
    </li>
    <li>
      <a href="../articles/model_pctl.html">Pan-cancer T cell lanscape</a>
    </li>
    <li>
      <a href="../articles/model_mtca.html">Mouse testicular cell atlas</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Map query dataset</li>
    <li>
      <a href="../articles/mapQuery_immune_RPL.html">Decidual immune microenvironment of RPL patients</a>
    </li>
    <li>
      <a href="../articles/mapQuery_ICB_BRCA.html">ICB responsiveness of BRCA</a>
    </li>
    <li>
      <a href="../articles/mapQuery_Zfp541-KO.html">Genetic perturbed germ cells (Zfp541-KO)</a>
    </li>
    <li>
      <a href="../articles/mapQuery_Ythdc2-KO.html">Genetic perturbed germ cells (Ythdc2-KO)</a>
    </li>
    <li>
      <a href="../articles/mapQuery_invitro_meiosis.html">In vitro induced meiosis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">MISC</li>
    <li>
      <a href="../articles/misc_Ythdc2-KO_pseudotime.html">Training the reference pseudotime model</a>
    </li>
    <li>
      <a href="../articles/misc_disco_pbmc_svm.html">Training the SVM model for cell type prediction</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/JarningGau/ProjectSVR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accurate identification of the tumor-infiltrated
T cell heterogeneity in cancer immunology study</h1>
            
            <h4 data-toc-skip class="date">Compiled: September 16,
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/JarningGau/ProjectSVR/blob/HEAD/vignettes/mapQuery_ICB_BRCA.Rmd" class="external-link"><code>vignettes/mapQuery_ICB_BRCA.Rmd</code></a></small>
      <div class="hidden name"><code>mapQuery_ICB_BRCA.Rmd</code></div>

    </div>

    
    
<p>In recent years, single-cell RNA-sequencing (scRNA-seq) enabled
unbiased exploration of T cell diversity in health, disease and response
to therapies at an unprecedented scale and resolution. However, a
comprehensive definition of T cell “reference” subtypes remains elusive
<a href="#refer-anchor-1"><sup>1</sup></a>.</p>
<p>Several groups try to integrate T cells in the tumor
micro-environment (TME) across different cancer types to build a
pan-cancer T cell landscape. For example, Zheng et al. depicted a
pan-cancer landscape of T cell heterogeneity in the TME and established
a baseline reference for future scRNA-seq studies associated with cancer
treatments <a href="#refer-anchor-2"><sup>2</sup></a>.</p>
<p>In this tutorial, we demonstrate the process of projecting query T
cells onto the pan-cancer T cell landscape (PTCL) to interpret the
disparities in CD8 T cell heterogeneity between responsive and
non-responsive patients with breast cancer who have undergone immune
checkpoint blockade (ICB) therapy <a href="#refer-anchor-3"><sup>3</sup></a>. This will be done using
ProjectSVR.</p>
<div class="section level2">
<h2 id="download-related-dataset">Download Related Dataset<a class="anchor" aria-label="anchor" href="#download-related-dataset"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/JarningGau/ProjectSVR" class="external-link">ProjectSVR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fl">3600</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"timeout"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reference model</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"models"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"models"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8350732/files/model.Zheng2021.CD8Tcell.rds"</span>, </span>
<span>              destfile <span class="op">=</span> <span class="st">"models/model.Zheng2021.CD8Tcell.rds"</span><span class="op">)</span></span>
<span><span class="co"># query data</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"query"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"query"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://zenodo.org/record/8350748/files/ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds"</span>, </span>
<span>              destfile <span class="op">=</span> <span class="st">"query/ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="map-query-to-reference">Map Query to Reference<a class="anchor" aria-label="anchor" href="#map-query-to-reference"></a>
</h2>
<div class="section level3">
<h3 id="reference-mapping">Reference mapping<a class="anchor" aria-label="anchor" href="#reference-mapping"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"models/model.Zheng2021.CD8Tcell.rds"</span><span class="op">)</span></span>
<span><span class="va">seu.q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"query/ICB_BRCA.1864-counts_cd8tcell_cohort1.seurat.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"E"</span> <span class="op">=</span> <span class="st">"Responders"</span>,</span>
<span>  <span class="st">"n/a"</span> <span class="op">=</span> <span class="st">"n/a"</span>,</span>
<span>  <span class="st">"NE"</span> <span class="op">=</span> <span class="st">"Non-Responders"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">seu.q</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">annotations</span><span class="op">[</span><span class="va">seu.q</span><span class="op">$</span><span class="va">expansion</span><span class="op">]</span></span>
<span></span>
<span><span class="va">seu.q</span> <span class="op">&lt;-</span> <span class="fu">ProjectSVR</span><span class="fu">::</span><span class="fu"><a href="../reference/MapQuery.html">MapQuery</a></span><span class="op">(</span><span class="va">seu.q</span>, reference <span class="op">=</span> <span class="va">reference</span>, </span>
<span>                              add.map.qual <span class="op">=</span> <span class="cn">T</span>, ncores <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu.q</span>, reduction <span class="op">=</span> <span class="st">"ref.umap"</span>, group.by <span class="op">=</span> <span class="st">"patient_id"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu.q</span>, reduction <span class="op">=</span> <span class="st">"ref.umap"</span>, group.by <span class="op">=</span> <span class="st">"cellSubType"</span>, </span>
<span>              label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggsci</span><span class="fu">::</span><span class="fu"><a href="https://nanx.me/ggsci/reference/scale_d3.html" class="external-link">scale_color_d3</a></span><span class="op">(</span><span class="st">"category20"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/reference_mapping-1.png" width="1152"></p>
</div>
<div class="section level3">
<h3 id="maping-quality">Maping quality<a class="anchor" aria-label="anchor" href="#maping-quality"></a>
</h3>
<p>Determine the threshold to distinguish the good and bad
projections.</p>
<p>QC plot for the reference mapping.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## cutoff by adjusted p value</span></span>
<span><span class="fu"><a href="../reference/MapQCPlot.html">MapQCPlot</a></span><span class="op">(</span><span class="va">seu.q</span>, p.adj.cutoff <span class="op">=</span> <span class="fl">1e-3</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/map_qc_plot-1.png" width="1344"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## or mean.knn.dist</span></span>
<span><span class="fu"><a href="../reference/MapQCPlot.html">MapQCPlot</a></span><span class="op">(</span><span class="va">seu.q</span>, map.q.cutoff <span class="op">=</span> <span class="fl">1.6</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/map_qc_plot-2.png" width="1344"></p>
<p>Visualize the projected query cells onto the reference atlas.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotProjection.html">PlotProjection</a></span><span class="op">(</span><span class="va">seu.q</span>, <span class="va">reference</span>, split.by <span class="op">=</span> <span class="st">"cellSubType"</span>, ref.color.by <span class="op">=</span> <span class="st">"cluster.name"</span>,</span>
<span>               ref.size <span class="op">=</span> <span class="fl">.5</span>, ref.alpha <span class="op">=</span> <span class="fl">.3</span>, query.size <span class="op">=</span> <span class="fl">1</span>, query.alpha <span class="op">=</span> <span class="fl">.5</span>, n.row <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/projection_plot-1.png" width="1440"></p>
</div>
<div class="section level3">
<h3 id="label-transfer">Label transfer<a class="anchor" aria-label="anchor" href="#label-transfer"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.q</span> <span class="op">&lt;-</span> <span class="fu">ProjectSVR</span><span class="fu">::</span><span class="fu"><a href="../reference/LabelTransfer.html">LabelTransfer</a></span><span class="op">(</span><span class="va">seu.q</span>, <span class="va">reference</span>, ref.label.col <span class="op">=</span> <span class="st">"cluster.name"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="differential-cell-population">Differential cell population<a class="anchor" aria-label="anchor" href="#differential-cell-population"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu.q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seu.q</span>, <span class="va">group</span> <span class="op">!=</span> <span class="st">"n/a"</span> <span class="op">&amp;</span> <span class="va">timepoint</span> <span class="op">==</span> <span class="st">"Pre"</span><span class="op">)</span></span>
<span><span class="va">seu.q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seu.q</span>, <span class="va">mean.knn.dist</span> <span class="op">&lt;</span> <span class="fl">1.6</span><span class="op">)</span></span>
<span><span class="va">seu.q</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">seu.q</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/PlotProjection.html">PlotProjection</a></span><span class="op">(</span><span class="va">seu.q</span>, <span class="va">reference</span>, split.by <span class="op">=</span> <span class="st">"group"</span>, ref.color.by <span class="op">=</span> <span class="st">"cluster.name"</span>,</span>
<span>               ref.size <span class="op">=</span> <span class="fl">.5</span>, ref.alpha <span class="op">=</span> <span class="fl">.3</span>, query.size <span class="op">=</span> <span class="fl">1</span>, query.alpha <span class="op">=</span> <span class="fl">.5</span>, n.row <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/unnamed-chunk-1-1.png" width="1056"></p>
<div class="section level3">
<h3 id="alluvia-plot">Alluvia plot<a class="anchor" aria-label="anchor" href="#alluvia-plot"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/AlluviaPlot.html">AlluviaPlot</a></span><span class="op">(</span><span class="va">seu.q</span><span class="op">@</span><span class="va">meta.data</span>, by <span class="op">=</span> <span class="st">"group"</span>, fill <span class="op">=</span> <span class="st">"knn.pred.celltype"</span>,</span>
<span>            colors <span class="op">=</span> <span class="va">reference</span><span class="op">$</span><span class="va">ref.cellmeta</span><span class="op">$</span><span class="va">colors</span>,</span>
<span>            bar.width <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/alluvia_plot-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="preference-analysis">Preference analysis<a class="anchor" aria-label="anchor" href="#preference-analysis"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GroupPreferencePlot.html">GroupPreferencePlot</a></span><span class="op">(</span><span class="va">seu.q</span><span class="op">@</span><span class="va">meta.data</span>, group.by <span class="op">=</span> <span class="st">"group"</span>, </span>
<span>                    preference.on <span class="op">=</span> <span class="st">"knn.pred.celltype"</span>, </span>
<span>                    column_names_rot <span class="op">=</span> <span class="fl">0</span>, column_names_centered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/preference_plot-1.png" width="576"></p>
<p>If <code>seu.q$knn.pred.celltype</code> was not a factor, then rows
were ordered by o/e value.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">seu.q</span><span class="op">@</span><span class="va">meta.data</span></span>
<span><span class="va">meta.data</span><span class="op">$</span><span class="va">knn.pred.celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">meta.data</span><span class="op">$</span><span class="va">knn.pred.celltype</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/GroupPreferencePlot.html">GroupPreferencePlot</a></span><span class="op">(</span><span class="va">meta.data</span>, group.by <span class="op">=</span> <span class="st">"group"</span>, </span>
<span>                    preference.on <span class="op">=</span> <span class="st">"knn.pred.celltype"</span>, </span>
<span>                    column_names_rot <span class="op">=</span> <span class="fl">0</span>, column_names_centered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/preference_plot2-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="wilcoxon-test">Wilcoxon test<a class="anchor" aria-label="anchor" href="#wilcoxon-test"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">da.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AbundanceTest.html">AbundanceTest</a></span><span class="op">(</span>cellmeta <span class="op">=</span> <span class="va">seu.q</span><span class="op">@</span><span class="va">meta.data</span>,</span>
<span>                         celltype.col <span class="op">=</span> <span class="st">"knn.pred.celltype"</span>,</span>
<span>                         sample.col <span class="op">=</span> <span class="st">"patient_id"</span>,</span>
<span>                         group.col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Volcano plot</span></span>
<span><span class="fu"><a href="../reference/VolcanoPlot.html">VolcanoPlot</a></span><span class="op">(</span><span class="va">da.test</span>, xlab <span class="op">=</span> <span class="st">"log2(R / NR)"</span>, ylab <span class="op">=</span> <span class="st">"Frequency"</span>,</span>
<span>            colors <span class="op">=</span> <span class="va">reference</span><span class="op">$</span><span class="va">ref.cellmeta</span><span class="op">$</span><span class="va">colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/wilcoxon_test-1.png" width="480"></p>
<p>Focus on the detailed information via <code><a href="../reference/BoxPlot.html">BoxPlot()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Box plot</span></span>
<span><span class="fu"><a href="../reference/BoxPlot.html">BoxPlot</a></span><span class="op">(</span>cellmeta <span class="op">=</span> <span class="va">seu.q</span><span class="op">@</span><span class="va">meta.data</span>, celltype.col <span class="op">=</span> <span class="st">"knn.pred.celltype"</span>,</span>
<span>        sample.col <span class="op">=</span> <span class="st">"patient_id"</span>, group.col <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>        celltypes.show <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8.c12(terminal Tex)"</span>, <span class="st">"CD8.c11(GZMK+ Tex)"</span><span class="op">)</span>,</span>
<span>        legend.ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NR"</span>, <span class="st">"R"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="mapQuery_ICB_BRCA_files/figure-html/box_plot-1.png" width="576"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.1.2 (2021-11-01)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] lubridate_1.9.2    forcats_1.0.0      stringr_1.5.0      dplyr_1.1.3       </span></span>
<span><span class="co">##  [5] purrr_1.0.2        readr_2.1.4        tidyr_1.3.0        tibble_3.2.1      </span></span>
<span><span class="co">##  [9] ggplot2_3.4.3      tidyverse_2.0.0    ProjectSVR_0.2.0   SeuratObject_4.1.3</span></span>
<span><span class="co">## [13] Seurat_4.3.0.1    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.3             spatstat.explore_3.2-3 reticulate_1.31       </span></span>
<span><span class="co">##   [4] tidyselect_1.2.0       mlr3learners_0.5.6     htmlwidgets_1.6.2     </span></span>
<span><span class="co">##   [7] BiocParallel_1.28.3    grid_4.1.2             Rtsne_0.16            </span></span>
<span><span class="co">##  [10] mlr3misc_0.12.0        munsell_0.5.0          codetools_0.2-18      </span></span>
<span><span class="co">##  [13] bbotk_0.7.2            ragg_1.2.5             ica_1.0-3             </span></span>
<span><span class="co">##  [16] future_1.33.0          miniUI_0.1.1.1         mlr3verse_0.2.8       </span></span>
<span><span class="co">##  [19] withr_2.5.0            spatstat.random_3.1-6  colorspace_2.1-0      </span></span>
<span><span class="co">##  [22] progressr_0.14.0       highr_0.10             knitr_1.43            </span></span>
<span><span class="co">##  [25] uuid_1.1-1             rstudioapi_0.15.0      stats4_4.1.2          </span></span>
<span><span class="co">##  [28] ROCR_1.0-11            robustbase_0.99-0      ggsignif_0.6.4        </span></span>
<span><span class="co">##  [31] tensor_1.5             listenv_0.9.0          labeling_0.4.3        </span></span>
<span><span class="co">##  [34] mlr3tuning_0.19.0      lgr_0.4.4              polyclip_1.10-4       </span></span>
<span><span class="co">##  [37] farver_2.1.1           rprojroot_2.0.3        parallelly_1.36.0     </span></span>
<span><span class="co">##  [40] vctrs_0.6.3            generics_0.1.3         xfun_0.40             </span></span>
<span><span class="co">##  [43] timechange_0.2.0       diptest_0.76-0         R6_2.5.1              </span></span>
<span><span class="co">##  [46] doParallel_1.0.17      clue_0.3-64            isoband_0.2.7         </span></span>
<span><span class="co">##  [49] flexmix_2.3-19         spatstat.utils_3.0-3   cachem_1.0.8          </span></span>
<span><span class="co">##  [52] promises_1.2.1         scales_1.2.1           nnet_7.3-17           </span></span>
<span><span class="co">##  [55] gtable_0.3.4           globals_0.16.2         mlr3hyperband_0.4.5   </span></span>
<span><span class="co">##  [58] goftest_1.2-3          mlr3mbo_0.2.1          rlang_1.1.1           </span></span>
<span><span class="co">##  [61] systemfonts_1.0.4      GlobalOptions_0.1.2    splines_4.1.2         </span></span>
<span><span class="co">##  [64] lazyeval_0.2.2         paradox_0.11.1         spatstat.geom_3.2-5   </span></span>
<span><span class="co">##  [67] checkmate_2.2.0        yaml_2.3.7             reshape2_1.4.4        </span></span>
<span><span class="co">##  [70] abind_1.4-5            mlr3_0.16.1            backports_1.4.1       </span></span>
<span><span class="co">##  [73] httpuv_1.6.11          tools_4.1.2            ellipsis_0.3.2        </span></span>
<span><span class="co">##  [76] jquerylib_0.1.4        RColorBrewer_1.1-3     BiocGenerics_0.40.0   </span></span>
<span><span class="co">##  [79] ggridges_0.5.4         Rcpp_1.0.11            plyr_1.8.8            </span></span>
<span><span class="co">##  [82] deldir_1.0-9           pbapply_1.7-2          GetoptLong_1.0.5      </span></span>
<span><span class="co">##  [85] cowplot_1.1.1          S4Vectors_0.32.4       zoo_1.8-12            </span></span>
<span><span class="co">##  [88] ggrepel_0.9.3          cluster_2.1.2          here_1.0.1            </span></span>
<span><span class="co">##  [91] fs_1.6.3               magrittr_2.0.3         data.table_1.14.8     </span></span>
<span><span class="co">##  [94] scattermore_1.2        circlize_0.4.15        lmtest_0.9-40         </span></span>
<span><span class="co">##  [97] RANN_2.6.1             fitdistrplus_1.1-11    matrixStats_1.0.0     </span></span>
<span><span class="co">## [100] hms_1.1.3              patchwork_1.1.3        mime_0.12             </span></span>
<span><span class="co">## [103] evaluate_0.21          xtable_1.8-4           mclust_6.0.0          </span></span>
<span><span class="co">## [106] IRanges_2.28.0         gridExtra_2.3          shape_1.4.6           </span></span>
<span><span class="co">## [109] UCell_1.3.1            compiler_4.1.2         mlr3cluster_0.1.8     </span></span>
<span><span class="co">## [112] KernSmooth_2.23-20     crayon_1.5.2           htmltools_0.5.6       </span></span>
<span><span class="co">## [115] tzdb_0.4.0             later_1.3.1            ComplexHeatmap_2.10.0 </span></span>
<span><span class="co">## [118] rappdirs_0.3.3         MASS_7.3-55            fpc_2.2-10            </span></span>
<span><span class="co">## [121] mlr3data_0.7.0         Matrix_1.6-1           cli_3.6.1             </span></span>
<span><span class="co">## [124] parallel_4.1.2         igraph_1.5.1           pkgconfig_2.0.3       </span></span>
<span><span class="co">## [127] pkgdown_2.0.7          sp_2.0-0               plotly_4.10.2         </span></span>
<span><span class="co">## [130] spatstat.sparse_3.0-2  foreach_1.5.2          bslib_0.5.1           </span></span>
<span><span class="co">## [133] mlr3fselect_0.11.0     digest_0.6.33          sctransform_0.3.5     </span></span>
<span><span class="co">## [136] RcppAnnoy_0.0.21       mlr3filters_0.7.1      spatstat.data_3.0-1   </span></span>
<span><span class="co">## [139] rmarkdown_2.24         leiden_0.4.3           uwot_0.1.16           </span></span>
<span><span class="co">## [142] kernlab_0.9-32         shiny_1.7.5            modeltools_0.2-23     </span></span>
<span><span class="co">## [145] rjson_0.2.21           lifecycle_1.0.3        nlme_3.1-155          </span></span>
<span><span class="co">## [148] jsonlite_1.8.7         mlr3tuningspaces_0.4.0 desc_1.4.2            </span></span>
<span><span class="co">## [151] viridisLite_0.4.2      fansi_1.0.4            pillar_1.9.0          </span></span>
<span><span class="co">## [154] ggsci_3.0.0            lattice_0.20-45        fastmap_1.1.1         </span></span>
<span><span class="co">## [157] httr_1.4.7             DEoptimR_1.1-2         survival_3.2-13       </span></span>
<span><span class="co">## [160] glue_1.6.2             mlr3viz_0.6.1          png_0.1-8             </span></span>
<span><span class="co">## [163] prabclus_2.3-2         iterators_1.0.14       spacefillr_0.3.2      </span></span>
<span><span class="co">## [166] class_7.3-20           stringi_1.7.12         sass_0.4.7            </span></span>
<span><span class="co">## [169] mlr3pipelines_0.5.0-1  palmerpenguins_0.1.1   textshaping_0.3.6     </span></span>
<span><span class="co">## [172] memoise_2.0.1          irlba_2.3.5.1          future.apply_1.11.0</span></span></code></pre>
</details>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refer-anchor-1">

</div>
<ul>
<li>[1] <a href="https://pubmed.ncbi.nlm.nih.gov/34017005/" class="external-link">Andreatta M,
et al. Interpretation of T cell states from single-cell transcriptomics
data using reference atlases. Nat Commun. 2021 May 20;12(1):2965. doi:
10.1038/s41467-021-23324-4. PMID: 34017005.</a>
</li>
</ul>
<div id="refer-anchor-2">

</div>
<ul>
<li>[2] <a href="https://pubmed.ncbi.nlm.nih.gov/34914499/" class="external-link">Zheng L, et
al. Pan-cancer single-cell landscape of tumor-infiltrating T cells.
Science. 2021 Dec 17;374(6574):abe6474. doi: 10.1126/science.abe6474.
Epub 2021 Dec 17. PMID: 34914499.</a>
</li>
</ul>
<div id="refer-anchor-3">

</div>
<ul>
<li>[3] <a href="https://pubmed.ncbi.nlm.nih.gov/33958794/" class="external-link">Bassez A, et
al. A single-cell map of intratumoral changes during anti-PD1 treatment
of patients with breast cancer. Nat Med. 2021 May;27(5):820-832. doi:
10.1038/s41591-021-01323-8. Epub 2021 May 6. PMID: 33958794.</a>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jianing Gao.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
